src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,169 @@
+        unlink(nodeToRemove);
+        size--;
+        T returnValue = nodeToRemove.item;
+        nodeToRemove.item = null;
+        return returnValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if ((object == current.item)
+                    || (object != null && object.equals(current.item))) {
+                unlink(current);
+                current.item = null;
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void unlink(Node<T> nodeToUnlink) {
+        if (nodeToUnlink.equals(head)) {
+            head = head.next;
+            nodeToUnlink.next = null;
+        } else if (nodeToUnlink.equals(tail)) {
+            tail = tail.prev;
+            nodeToUnlink.prev = null;
+        } else {
+            nodeToUnlink.prev.next = nodeToUnlink.next;
+            nodeToUnlink.next.prev = nodeToUnlink.prev;
+        }
+    }
+
+    private void addToHead(Node<T> node) {
+        if (head == null) {
+            head = node;
+            tail = node;
+        } else {
+            head.prev = node;
+            node.next = head;
+            head = node;
+        }
+    }
+
+    private void addByIndex(Node<T> node, int index) {
+        if (index == size) {
+            tail.next = node;
+            node.prev = tail;
+            tail = node;
+        } else {
+            Node<T> nodeByIndex = getNodeByIndex(index);
+            node.prev = nodeByIndex.prev;
+            nodeByIndex.prev.next = node;
+            nodeByIndex.prev = node;
+            node.next = nodeByIndex;
+        }
+    }
+
+    private void throwException() {
+        throw new IndexOutOfBoundsException(EXCEPTION_MESSAGE);
+    }
+
+    private void validateIndex(int index) {
+        if (index < 0 || index > size) {
+            throwException();
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(EXCEPTION_MESSAGE);
+        }
+        Node<T> resultNode;
+        if (index < size / 2) {
+            resultNode = head;
+            for (int i = 0; i < index; i++) {
+                resultNode = resultNode.next;
+            }
+        } else {
+            resultNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                resultNode = resultNode.prev;
+            }
+        }
+        return resultNode;
+    }
+
+    private static class Node<T> {
+        private Node<T> prev;
+        private Node<T> next;
+        private T item;
+
+        public Node(Node<T> previousNode, T item, Node<T> nextNode) {
```

Comment:
for better readability put that class declaration on top

Reply:
It's a very discussable point. As I know, depends on the team codestyle where the inner class declaration should be located. I'm guided by the logic that since it's private, it should be declared after all the public stuff

Reply:
it does actually depend on the team, but I lean towards the end of the file since it is private